patches/api/0420-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,746 @@
+@@ -0,0 +1,10 @@
++package io.papermc.paper.command.brigadier;
++
++import java.util.ServiceLoader;
++
++final class MessageComponentSerializerHolder {
++
++    public static final MessageComponentSerializer PROVIDER = ServiceLoader.load(MessageComponentSerializer.class)
++        .findFirst()
++        .orElseThrow();
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..14ff48fb772a9f0ca9a3d37f8a600a39d1d14a8c
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
+@@ -0,0 +1,24 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import io.papermc.paper.command.brigadier.CommandSourceStack;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Represents an argument which requires
++ * the context of a command source stack to be fully resolved.
++ * @param <T> type
++ */
++@ApiStatus.Experimental
++public interface ArgumentResolver<T> {
++
++    /**
++     * Resolves the argument with the given
++     * command source stack.
++     * @param sourceStack source stack
++     * @return resolved
++     */
++    @NotNull
++    T resolve(@NotNull CommandSourceStack sourceStack) throws CommandSyntaxException;
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..50a54e913a2aad1dcf117814de353e9b6fcadcc6
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java
+@@ -0,0 +1,87 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.mojang.brigadier.StringReader;
++import com.mojang.brigadier.arguments.ArgumentType;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++
++/**
++ * An argument type that wraps around another argument.
++ * This argument is special in that the argument type that is wrapped will
++ * be sent to the client, not the wrapper itself.
++ *
++ * @param <T> type
++ * @param <B> base wrapping type
++ */
++@ApiStatus.Experimental
++public abstract class CustomArgumentType<T, B> implements ArgumentType<T> {
++
++    @NotNull
++    private final ArgumentType<B> vanillaPrimitiveType;
++
++    public CustomArgumentType(@NotNull ArgumentType<B> primitive) {
++        VanillaArguments.PROVIDER.validatePrimitiveType(primitive);
++        this.vanillaPrimitiveType = primitive;
++    }
++
++    /**
++     * Parses the argument using the wrapped argument type then converting
```

Comment:
I wouldn't even mention "wraping" or "primitive" in this javadoc, that's not really important right? just converts the input to your custom type